strawberry/schema/schema_converter.py

```diff
@@ -721,8 +721,18 @@ def validate_same_type_definition(
                 # may be referencing the same actual type
                 if isinstance(type1, LazyType):
                     type1 = type1.resolve_type()
+                elif isinstance(type1, StrawberryOptional) and isinstance(
+                    type1.of_type, LazyType
+                ):
+                    type1.of_type = type1.of_type.resolve_type()
+
                 if isinstance(type2, LazyType):
                     type2 = type2.resolve_type()
+                elif isinstance(type2, StrawberryOptional) and isinstance(
+                    type2.of_type, LazyType
+                ):
+                    type2.of_type = type2.of_type.resolve_type()
```

Comment:
do we need to do this for lists as well?

Reply:
Yeah 🙁 I believe also for `StrawberryUnion`. I think the right way to go is to define `__eq__` on `TypeDefinition` so then in this method we can just compare equality on the types directly. The derived type classes can then implement their own behavior of how equality should be checked. For now I think this bit of code will get out of hand if we implement all cases. It's meant more of as a quick solution for @gauthamkamath issue. If you think it should be done in this PR I can work on implementing `__eq__` tonight. Otherwise I'll do it in a follow-up PR

Reply:
let's do it in another PR!